libsupport/src/Plugin.cpp

```diff
@@ -0,0 +1,187 @@
+                           .so_path = path.string(),
+                           .so_handle = handle,
+                       });
+  }
+
+  void LoadAllPlugins() {
+    if (!plugins_.empty()) {
+      return;
+    }
+
+    auto path = katana::GetPluginPath();
+    for (const auto& p : path) {
+      KATANA_LOG_VERBOSE("Loading plugins from: {}", p);
+      for (const boost::filesystem::directory_entry& entry :
+           boost::filesystem::directory_iterator(
+               p, boost::filesystem::directory_options::none)) {
+        // TODO(amp): This only supports Linux.
+        if (entry.path().extension() == ".so") {
+          boost::system::error_code err{};
+          boost::filesystem::path file = canonical(entry.path(), err);
+          if (err) {
+            KATANA_LOG_WARN(
+                "skipping plugin {}: could not canonicalize plugin path: {}",
+                entry.path(), err);
+          }
+          if (is_regular_file(file, err)) {
+            LoadPlugin(file);
+          } else {
+            KATANA_LOG_WARN("skipping plugin {}: not a regular file", file);
+          }
+          if (err) {
+            KATANA_LOG_WARN(
+                "skipping plugin {}: could not stat plugin: {}", file, err);
+          }
+        }
+      }
+    }
+  }
+
+  std::vector<katana::Plugin> plugins() const {
+    std::vector<katana::Plugin> ret(plugins_.size());
+
+    transform(plugins_.begin(), plugins_.end(), ret.begin(), [&](auto p) {
+      return p.second;
+    });
+
+    return ret;
+  }
+
+  void FinalizePlugins() {
+    for (const auto& entry : plugins_) {
+      const katana::Plugin& plugin = entry.second;
+      plugin.metadata->finalize();
+      dlclose(plugin.so_handle);
+    }
+    plugins_.clear();
+  }
+};
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+PluginLoader plugin_loader;
+
+}  // namespace
+
+std::vector<boost::filesystem::path>
+katana::GetPluginPath() {
+  std::vector<boost::filesystem::path> ret;
+  bool found = false;
+
+  // 1) User provided paths from KATANA_PLUGIN_PATH overriding all other paths
+  const char* env_var_str = std::getenv("KATANA_PLUGIN_PATH");
+  if (env_var_str) {
+    std::string env_var_path(env_var_str);
+    for (const auto& s : katana::SplitView(env_var_path, ":")) {
+      if (!s.empty()) {
+        ret.emplace_back(std::string(s));
+        found = true;
+      }
+    }
+  } else {
+    // 2) Path relative to the shared object
+    Dl_info dl_info;
+    // Get the path of the shared object containing this function and compute
+    // paths relative to it.
+    if (dladdr(reinterpret_cast<void*>(&GetPluginPath), &dl_info)) {
+      boost::filesystem::path this_so_filename(dl_info.dli_fname);
+      boost::system::error_code err{};
+      this_so_filename = canonical(this_so_filename, err);
+      if (!err) {
+        // 2a) installed shared objects
+        auto path = canonical(
+            this_so_filename.parent_path() / "katana" / "plugins", err);
```

Comment:
can we dump plugins to CMAKE_BINARY_PATH or something so we don't need to hardcode the relation of different build directories?

Reply:
Yea. This is a hack. Putting them in `CMAKE_BINARY_PATH` wouldn't help since this loading doesn't use `LD_LIBRARY_PATH` or rpaths. I did this because it means I don't need to change all the launchers, but I could remove this and instead set `KATANA_PLUGIN_PATH` in the ctest config and is `python_env.sh`. What's your preference?

Reply:
I don't think `python_env.sh` would work that well because we'd have to train people to use that always.

Nothing obvious comes up to me for cleaning this up.

Though, perhaps we should be more strict about the paths we check: (1) if KATANA_PLUGIN_PATH is set use it and no other; otherwise (2) if installed plugin directory exists, use it and no other; finally (3) try only build-tree based paths.

Reply:
I actually already do 1 and 2. I will add 3 as conditional step. I will also add a WARNING that is printed when build tree plugins are used, so people don't get surprised by it.

And yes, I just realized the issue. People obviously run native programs directly. Don't know how I forgot that.